traffic_ops/traffic_ops_golang/invalidationjobs/invalidationjobs.go

```diff
@@ -0,0 +1,928 @@
+			return
+		}
+	} else {
+		w.Header().Set(http.CanonicalHeaderKey("accept"), strings.Join([]string{tc.ApplicationJson, "application/x-www-form-urlencoded", "multipart/form-data"}, ";"))
+		err := fmt.Errorf("unsupported content-type: %s", cType)
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusUnsupportedMediaType, err, err)
+		return
+	}
+
+	if err := input.Validate(); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	if job.StartTime.Before(time.Now()) {
+		userErr = errors.New("Cannot modify a job that has already started!")
+		errCode = http.StatusMethodNotAllowed
+		w.Header().Set(http.CanonicalHeaderKey("allow"), "GET,HEAD,DELETE")
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.DeliveryService != *input.DeliveryService {
+		userErr = errors.New("Cannot change 'deliveryService' of existing invalidation job!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.CreatedBy != *input.CreatedBy {
+		userErr = errors.New("Cannot change 'createdBy' of existing invalidation jobs!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.ID != *input.ID {
+		userErr = errors.New("Cannot change an invalidation job 'id'!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+
+	row = inf.Tx.Tx.QueryRow(updateQuery,
+		input.AssetURL,
+		input.Keyword,
+		input.Parameters,
+		input.StartTime.Time,
+		*job.ID)
+	err = row.Scan(&job.AssetURL,
+		&job.CreatedBy,
+		&job.DeliveryService,
+		&job.ID,
+		&job.Keyword,
+		&job.Parameters,
+		&job.StartTime)
+	if err != nil {
+		sysErr = fmt.Errorf("Updating a job: %v", err)
+		errCode = http.StatusInternalServerError
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, nil, sysErr)
+		return
+	}
+
+	if err = setRevalFlags(*job.DeliveryService, inf.Tx.Tx); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("Setting reval flags: %v", err))
+		return
+	}
+
+	response := apiResponse{
+		[]tc.Alert{
+			tc.Alert{
+				"Content invalidation job updated",
+				tc.SuccessLevel.String(),
+			},
+		},
+		job,
+	}
+
+	resp, err := json.Marshal(response)
+	if err != nil {
+		sysErr = fmt.Errorf("encoding response: %v", err)
+		errCode = http.StatusInternalServerError
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, nil, sysErr)
+	} else {
+		w.Header().Set(http.CanonicalHeaderKey("content-type"), tc.ApplicationJson)
+		w.Write(resp)
+		w.Write([]byte{'\n'})
+	}
+
+	api.CreateChangeLogRawTx(api.ApiChange, api.Updated+"content invalidation job: #"+strconv.FormatUint(*job.ID, 10), inf.User, inf.Tx.Tx)
+}
+
+// Used by DELETE requests to `/jobs`, deletes an existing content invalidation job
+func Delete(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"id"}, []string{"id"})
```

Comment:
if a path always requires a query parameter should we make it part of the path?

Reply:
Mainly because it's less endpoints to document. But also because it's not necessarily true that the query parameter will always be required, for example suppose we want to DELETE all jobs where the start time is at least a week ago and the time to live is less than a week.